<a href="https://colab.research.google.com/github/samkemboi12/Smart_phone_credit_scoring_model/blob/main/Credit_Score_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from glob import glob

In [2]:
!git clone https://github.com/gregory-bot/MoPhones-Case-Study


Cloning into 'MoPhones-Case-Study'...
remote: Enumerating objects: 31, done.
remote: Counting objects: 100% (31/31), done.
remote: Compressing objects: 100% (24/24), done.
remote: Total 31 (delta 12), reused 24 (delta 7), pack-reused 0 (from 0)
Receiving objects: 100% (31/31), 13.87 MiB | 19.27 MiB/s, done.
Resolving deltas: 100% (12/12), done.


In [3]:
cd MoPhones-Case-Study/'Credit Data'

/content/MoPhones-Case-Study/Credit Data


In [4]:
df = pd.read_excel("Credit Data Definitions.xlsx")
df

,Column Name,Definition
0,LOAN_ID,Unique identifier for the loan.
1,DATE,Reporting (snapshot) date on which all loan me...
2,CUSTOMER_AGE,Number of days since the loan sale date (SALE_...
3,TOTAL_PAID,Cumulative amount paid by the customer up to t...
4,TOTAL_DUE_TODAY,Total amount the customer is expected to have ...
5,BALANCE,Outstanding loan balance after accounting for ...
6,DAYS_PAST_DUE,Number of days the customer is behind on payme...
7,CLOSING_BALANCE,Final balance for reporting; set to zero if th...
8,ADVANCE,Amount the customer has paid ahead of schedule...
9,BALANCE_DUE_TO_DATE,Net payment position of the customer; positive...


In [5]:
cols_to_keep = [
    "LOAN_ID",
    "DATE",
    "SALE_DATE",

    "CUSTOMER_AGE",
    "ADVANCE",

    "TOTAL_PAID",
    "TOTAL_DUE_TODAY",

    "INITIAL_PAY",
    "WEEKLY_RATE",
    "EXPECTED_PAYMENT",
    "DEPOSIT",

    "DISCOUNT",
    "CREDIT_CHECK_DONE",
    "BALANCE",
    "CLOSING_BALANCE",
    "BALANCE_DUE_TO_DATE",
    "ARREARS",

    "PAYMENT",
    "FIRST_PAYMENT",
    "FIRST_EXPECTED_PAYMENT",

    "ACCOUNT_STATUS_L1",
    "ACCOUNT_STATUS_L2",

    "OVERPAYMENT_AMOUNT",

    "CREDIT_EXPIRY",
    "NEXT_INVOICE_DATE",
    "MAX_PAYMENT_DATE",
    "DAYS_PAST_DUE",
    "TOTAL_PAID_WITH_ADJUSTMENTS_15D"
]
credit1 = pd.read_csv("Credit Data - 30-12-2025.csv")
credit1 = credit1[cols_to_keep]
credit2 = pd.read_csv("Credit Data - 30-09-2025.csv")
credit2 = credit2[cols_to_keep]
credit3 = pd.read_csv("Credit Data - 30-06-2025.csv")
credit3 = credit3[cols_to_keep]
credit4 = pd.read_csv("Credit Data - 30-03-2025.csv")
credit4 = credit4[cols_to_keep]
credit5 = pd.read_csv("Credit Data - 01-01-2025.csv")
credit5 = credit5[cols_to_keep]

In [6]:
for i, df in enumerate([credit1, credit2, credit3, credit4, credit5], start=1):
    print(f"\n===== DATASET {i} =====")
    print(df.info())


===== DATASET 1 =====
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20742 entries, 0 to 20741
Data columns (total 28 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   LOAN_ID                          20742 non-null  object 
 1   DATE                             20742 non-null  object 
 2   SALE_DATE                        20742 non-null  object 
 3   CUSTOMER_AGE                     20742 non-null  int64  
 4   ADVANCE                          20742 non-null  float64
 5   TOTAL_PAID                       20742 non-null  int64  
 6   TOTAL_DUE_TODAY                  20736 non-null  float64
 7   INITIAL_PAY                      20742 non-null  int64  
 8   WEEKLY_RATE                      20740 non-null  float64
 9   EXPECTED_PAYMENT                 20742 non-null  int64  
 10  DEPOSIT                          20740 non-null  float64
 11  DISCOUNT                         20742 non-null  float64


In [7]:
dfs = [credit1, credit2, credit3, credit4, credit5]
for df in dfs:
    df["ADVANCE"] = df["ADVANCE"].astype(float)
    df["DISCOUNT"] = df["DISCOUNT"].astype(float)

In [8]:
credit = pd.concat(dfs, ignore_index=True)

In [9]:
credit.info()
credit.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71456 entries, 0 to 71455
Data columns (total 28 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   LOAN_ID                          71456 non-null  object 
 1   DATE                             71456 non-null  object 
 2   SALE_DATE                        71456 non-null  object 
 3   CUSTOMER_AGE                     71456 non-null  int64  
 4   ADVANCE                          71456 non-null  float64
 5   TOTAL_PAID                       71456 non-null  int64  
 6   TOTAL_DUE_TODAY                  71428 non-null  float64
 7   INITIAL_PAY                      71456 non-null  int64  
 8   WEEKLY_RATE                      71448 non-null  float64
 9   EXPECTED_PAYMENT                 71455 non-null  float64
 10  DEPOSIT                          71448 non-null  float64
 11  DISCOUNT                         71456 non-null  float64
 12  CREDIT_CHECK_DONE 

,0
LOAN_ID,0
DATE,0
SALE_DATE,0
CUSTOMER_AGE,0
ADVANCE,0
TOTAL_PAID,0
TOTAL_DUE_TODAY,28
INITIAL_PAY,0
WEEKLY_RATE,8
EXPECTED_PAYMENT,1


In [10]:
for col in ["WEEKLY_RATE", "DEPOSIT", "BALANCE"]:
    credit[col] = credit[col].fillna(credit[col].median())

In [11]:
credit.head(15)

,LOAN_ID,DATE,SALE_DATE,CUSTOMER_AGE,ADVANCE,TOTAL_PAID,TOTAL_DUE_TODAY,INITIAL_PAY,WEEKLY_RATE,EXPECTED_PAYMENT,...,FIRST_PAYMENT,FIRST_EXPECTED_PAYMENT,ACCOUNT_STATUS_L1,ACCOUNT_STATUS_L2,OVERPAYMENT_AMOUNT,CREDIT_EXPIRY,NEXT_INVOICE_DATE,MAX_PAYMENT_DATE,DAYS_PAST_DUE,TOTAL_PAID_WITH_ADJUSTMENTS_15D
0,reczZiie22UfW6jzq,12/30/2025,10/15/2024,441,0.0,68959,68959.0,8779,1180.0,0.0,...,0,0,Paid Off Loan,Paid Off,0.0,10/14/2025,1/6/2026,12/23/2025,0,68959
1,recvDazucywbN0XCG,12/30/2025,8/22/2025,130,0.0,11399,11399.0,3299,450.0,0.0,...,0,0,Existing Active,Active,0.0,1/2/2026,1/2/2026,12/25/2025,0,11399
2,recEc4UFc7rXjdNqc,12/30/2025,7/31/2023,883,0.0,137280,151459.0,27459,2480.0,0.0,...,0,0,Paid Off Loan,Paid Off,0.0,6/10/2024,1/5/2026,1/28/2024,0,137280
3,rec1a55A7ospM1pyY,12/30/2025,11/28/2025,32,0.0,11679,11679.0,6999,1170.0,0.0,...,0,0,Existing Active,Active,0.0,1/2/2026,1/2/2026,12/26/2025,0,11679
4,recqDWHPzwZrSsJii,12/30/2025,8/4/2024,513,0.0,44944,45039.0,5769,770.0,0.0,...,0,0,Paid Off Loan,Paid Off,0.0,7/27/2025,1/4/2026,7/27/2025,0,44944
5,recGbe6M7ZKV99teG,12/30/2025,12/28/2024,367,0.0,20589,82419.0,10509,1410.0,0.0,...,0,0,Write Off,PAR 30,0.0,2/22/2025,1/3/2026,2/23/2025,311,20589
6,recnB2BUhZsEsHXPu,12/30/2025,10/5/2024,451,0.0,10399,81799.0,10399,1400.0,0.0,...,0,0,First Payment Default,FPD,0.0,10/12/2024,1/3/2026,10/5/2024,444,10399
7,recZl6QI7J9j3WtM1,12/30/2025,6/21/2025,192,140.0,50229,50089.0,10399,1470.0,0.0,...,0,0,Existing Active,Active,0.0,1/3/2026,1/3/2026,12/27/2025,0,50229
8,rec8HFmF0yfh6LION,12/30/2025,6/4/2025,209,0.0,22999,30649.0,5999,850.0,0.0,...,0,0,Blocked 61-90,PAR 30,0.0,10/29/2025,12/31/2025,10/22/2025,62,22999
9,recUQwHhQrfDHkFb0,12/30/2025,11/19/2024,406,0.0,95879,95879.0,12239,1640.0,0.0,...,0,0,Paid Off Loan,Paid Off,0.0,11/18/2025,1/6/2026,12/18/2025,0,95879


In [12]:
credit["TOTAL_DUE_TODAY"] = credit["TOTAL_DUE_TODAY"].fillna(0)
credit["BALANCE_DUE_TO_DATE"] = credit["BALANCE_DUE_TO_DATE"].fillna(0)

In [13]:
date_cols = ["DATE", "SALE_DATE", "MAX_PAYMENT_DATE", "NEXT_INVOICE_DATE", "CREDIT_EXPIRY"]
for col in date_cols:
    credit[col] = pd.to_datetime(credit[col], dayfirst=False)

In [14]:
credit.duplicated(subset=["LOAN_ID", "DATE"]).sum()

np.int64(0)

In [15]:
credit["LOAN_ID"].nunique()

20742

In [16]:
(credit.isna().mean()*100).sort_values(ascending=False)

,0
MAX_PAYMENT_DATE,1.048197
CLOSING_BALANCE,0.011196
EXPECTED_PAYMENT,0.001399
LOAN_ID,0.000000
CUSTOMER_AGE,0.000000
ADVANCE,0.000000
TOTAL_DUE_TODAY,0.000000
TOTAL_PAID,0.000000
INITIAL_PAY,0.000000
WEEKLY_RATE,0.000000


In [17]:
credit = credit.drop(columns=["MAX_PAYMENT_DATE"])

In [18]:
credit["EXPECTED_PAYMENT"] = pd.to_numeric(credit["EXPECTED_PAYMENT"], errors="coerce")
credit["CLOSING_BALANCE"] = pd.to_numeric(credit["CLOSING_BALANCE"], errors="coerce")

In [19]:
credit.dtypes

,0
LOAN_ID,object
DATE,datetime64[ns]
SALE_DATE,datetime64[ns]
CUSTOMER_AGE,int64
ADVANCE,float64
TOTAL_PAID,int64
TOTAL_DUE_TODAY,float64
INITIAL_PAY,int64
WEEKLY_RATE,float64
EXPECTED_PAYMENT,float64


In [20]:
credit.skew(numeric_only=True)

,0
CUSTOMER_AGE,0.658936
ADVANCE,12.128690
TOTAL_PAID,1.476468
TOTAL_DUE_TODAY,1.253613
INITIAL_PAY,3.160391
WEEKLY_RATE,2.462891
EXPECTED_PAYMENT,25.293212
DEPOSIT,2.639862
DISCOUNT,11.158699
BALANCE,0.884952


In [21]:
high_skew_cols = [
    "OVERPAYMENT_AMOUNT",
    "PAYMENT",
    "FIRST_PAYMENT",
    "EXPECTED_PAYMENT",
    "FIRST_EXPECTED_PAYMENT"
]

In [22]:
credit.corr(numeric_only=True)

,CUSTOMER_AGE,ADVANCE,TOTAL_PAID,TOTAL_DUE_TODAY,INITIAL_PAY,WEEKLY_RATE,EXPECTED_PAYMENT,DEPOSIT,DISCOUNT,BALANCE,CLOSING_BALANCE,BALANCE_DUE_TO_DATE,ARREARS,PAYMENT,FIRST_PAYMENT,FIRST_EXPECTED_PAYMENT,OVERPAYMENT_AMOUNT,DAYS_PAST_DUE,TOTAL_PAID_WITH_ADJUSTMENTS_15D
CUSTOMER_AGE,1.000000,-0.098030,0.314299,0.523488,-0.098786,-0.096299,-0.081166,-0.081766,0.017479,-0.370914,-0.276207,-0.323188,0.317807,-0.093350,-0.054129,-0.054380,0.057337,0.537470,0.301079
ADVANCE,-0.098030,1.000000,0.169474,-0.002510,0.081436,0.112501,-0.009916,0.148060,0.365199,-0.063432,-0.082393,0.216346,-0.103765,0.038673,-0.003247,-0.003292,0.239979,-0.100761,0.175768
TOTAL_PAID,0.314299,0.169474,1.000000,0.691907,0.391089,0.508090,-0.030853,0.493725,0.178322,-0.426984,-0.442051,0.296220,-0.282002,0.000726,-0.034162,-0.034477,0.093520,-0.322013,0.993818
TOTAL_DUE_TODAY,0.523488,-0.002510,0.691907,1.000000,0.378758,0.637295,0.004238,0.578934,0.123098,0.041111,0.058640,-0.476728,0.485363,-0.044437,-0.047698,-0.048001,0.042688,0.282688,0.686900
INITIAL_PAY,-0.098786,0.081436,0.391089,0.378758,1.000000,0.621427,0.040063,0.717549,0.074562,0.124455,0.099297,-0.016603,0.026412,0.033441,0.002531,0.002303,0.007264,-0.129102,0.396515
WEEKLY_RATE,-0.096299,0.112501,0.508090,0.637295,0.621427,1.000000,0.087544,0.830034,0.117307,0.445244,0.365516,-0.208422,0.225447,0.045649,-0.000285,-0.000428,0.009880,-0.054574,0.517705
EXPECTED_PAYMENT,-0.081166,-0.009916,-0.030853,0.004238,0.040063,0.087544,1.000000,0.083063,-0.026642,0.135142,0.111789,-0.040753,0.040360,0.710834,0.654742,0.654537,-0.009228,0.011153,-0.029020
DEPOSIT,-0.081766,0.148060,0.493725,0.578934,0.717549,0.830034,0.083063,1.000000,0.142782,0.432938,0.363628,-0.140355,0.160253,0.047877,0.000716,0.000470,0.021886,-0.054622,0.510866
DISCOUNT,0.017479,0.365199,0.178322,0.123098,0.074562,0.117307,-0.026642,0.142782,1.000000,-0.144662,-0.124295,0.063517,-0.022112,-0.002820,-0.004762,-0.004784,-0.005100,-0.073944,0.185487
BALANCE,-0.370914,-0.063432,-0.426984,0.041111,0.124455,0.445244,0.135142,0.432938,-0.144662,1.000000,0.915445,-0.555426,0.558430,0.048934,0.033276,0.033423,-0.049246,0.301988,-0.414245


In [23]:
high_corr_pairs = [
    ("BALANCE", "CLOSING_BALANCE"),
    ("BALANCE_DUE_TO_DATE", "ARREARS"),
    ("FIRST_PAYMENT", "FIRST_EXPECTED_PAYMENT"),
    ("TOTAL_PAID", "TOTAL_PAID_WITH_ADJUSTMENTS_15D")
]

In [24]:
import os
cust_file = os.path.join("../", "Sales and Customer Data.xlsx")
gender = pd.read_excel(cust_file, sheet_name="Gender")
gender.columns = gender.columns.str.strip()
gender = gender.dropna(subset=["Loan Id"])
gender = gender.drop_duplicates(subset=["Loan Id"])

In [25]:
dob = pd.read_excel(cust_file, sheet_name="DOB")
dob.columns = dob.columns.str.strip()
dob = dob.dropna(subset=["Loan Id"])
dob["date_of_birth"] = pd.to_datetime(dob["date_of_birth"], errors="coerce", utc=True)
dob["date_of_birth"] = dob["date_of_birth"].dt.tz_localize(None)
dob = dob.drop_duplicates(subset=["Loan Id"])

In [26]:
income = pd.read_excel(cust_file, sheet_name="Income Level")
income.columns = income.columns.str.strip()
income = income.dropna(subset=["Loan Id"])
income = income.drop_duplicates(subset=["Loan Id"])

In [27]:
nps_file = os.path.join("../", "NPS Data.xlsx")
nps = pd.read_excel(nps_file)
nps.columns = nps.columns.str.strip()
nps = nps.rename(columns={
    "Using a scale from 0 (not likely) to 10 (very likely), how likely are you to recommend MoPhones to friends or family?": "NPS_SCORE",
    "What is the main reason for your score?": "NPS_REASON",
    "What is one thing we could do to improve your experience with us?": "IMPROVEMENT",
    "Are you happy with the quality and performance of your MoPhones device?": "HAPPY_DEVICE",
    "Are you happy with the service and support provided by MoPhones?": "HAPPY_SERVICE",
})
nps["NPS_SCORE"] = pd.to_numeric(nps["NPS_SCORE"], errors="coerce")
nps = nps.drop_duplicates(subset=["Loan Id"])
nps = nps.dropna(subset=["Loan Id"])

In [28]:
clean_credit_snapshot = credit.copy()
clean_customer_gender = gender.copy()
clean_customer_dob = dob.copy()
clean_customer_income = income.copy()
clean_nps_responses = nps.copy()

In [29]:
print(clean_credit_snapshot.shape)
print(clean_credit_snapshot.isnull().sum().sort_values(ascending=False).head(10))
print(clean_credit_snapshot.duplicated(subset=["LOAN_ID", "DATE"]).sum())

(71456, 27)
CLOSING_BALANCE     8
EXPECTED_PAYMENT    1
DATE                0
SALE_DATE           0
CUSTOMER_AGE        0
TOTAL_PAID          0
ADVANCE             0
TOTAL_DUE_TODAY     0
INITIAL_PAY         0
WEEKLY_RATE         0
dtype: int64
0


In [30]:
print(clean_customer_gender.dtypes)
print(clean_customer_dob.dtypes)
print(clean_customer_income.dtypes)
print(clean_nps_responses.dtypes)

Loan Id        object
Citizenship    object
Gender         object
dtype: object
_id                      object
provider                 object
date_of_birth    datetime64[ns]
Loan Id                  object
createdAt UTC            object
dtype: object
Loan Id                         object
Duration                       float64
Received                       float64
Persons Received From Total    float64
Banks Received                 float64
Paybills Received Others       float64
dtype: object
Submission ID                                                                                           object
Respondent ID                                                                                           object
Submitted at                                                                                    datetime64[ns]
Loan Id                                                                                                 object
NPS_SCORE                                             

In [31]:
def clean_loan_id(df):
    df["Loan Id"] = (
        df["Loan Id"]
        .astype(str)
        .str.strip()
        .str.upper()
    )
    return df
clean_customer_gender = clean_loan_id(clean_customer_gender)
clean_customer_dob = clean_loan_id(clean_customer_dob)
clean_customer_income = clean_loan_id(clean_customer_income)
clean_nps_responses = clean_loan_id(clean_nps_responses)
clean_customer_gender = clean_customer_gender.rename(columns={"Loan Id": "LOAN_ID"})
clean_customer_dob = clean_customer_dob.rename(columns={"Loan Id": "LOAN_ID"})
clean_customer_income = clean_customer_income.rename(columns={"Loan Id": "LOAN_ID"})
clean_nps_responses = clean_nps_responses.rename(columns={"Loan Id": "LOAN_ID"})

In [32]:
clean_nps_responses["Submitted at"] = pd.to_datetime(
    clean_nps_responses["Submitted at"],
    errors="coerce"
)
clean_customer_dob["createdAt UTC"] = pd.to_datetime(
    clean_customer_dob["createdAt UTC"],
    errors="coerce"
)

In [33]:
clean_credit_snapshot.to_csv("credit_cleaned.csv", index=False)
clean_customer_gender
clean_customer_dob
clean_customer_income
clean_nps_responses

,Submission ID,Respondent ID,Submitted at,LOAN_ID,NPS_SCORE,NPS_REASON,IMPROVEMENT,HAPPY_DEVICE,HAPPY_SERVICE,Have you ever experienced a delay in your payment reflecting in your Mophones account?,Have you ever had difficulty getting assistance from MoPhones customer support when needed?,(If Yes) – Please describe the challenge you faced and how we can improve your experience.,Have you experienced any battery-related issues with your MoPhones device?,Have you used the MoPhones app (MoApp) to manage your account or make payments?,Which communication channel do you prefer when contacting MoPhones for inquiries or support?,Have you ever had your phone lock despite making a payment on time?,Any other Feedback?
0,BzK4Q11,k4z1Y6,2025-04-22 15:15:00,REC2FCXHHQXW6AKRU,6.0,NaN,NaN,Yes,Yes,Yes,Yes,Unlocking my phone,No,"Yes, and I am satisfied with the MoApp",MoPhones App,Yes,NaN
1,b59V45Z,AY9GMk,2025-04-22 15:18:00,RECQCLWBPCEP18WDW,10.0,NaN,NaN,Yes,Yes,No,No,Response team always on time,No,"Yes, and I am satisfied with the MoApp",Free SMS – 25044,No,Would like to upgrade once I'm done with this ...
2,GxKvYKO,eGz8Ek,2025-04-22 15:31:00,RECEI9UEU0E5F9HUG,5.0,NaN,NaN,No,No,No,Yes,I bought a Samsung A42 which has fingerprint l...,No,"Yes, and I am satisfied with the MoApp",Phone Call – +254 728 444 442 or +254 709 924 404,Yes,Good service and efficient response
3,A7KP49z,LjOx7j,2025-04-22 15:40:00,RECOSDWAJVWZAKGLV,1.0,NaN,NaN,No,No,Yes,Yes,Delays,Yes,"Yes, but I am not satisfied with the MoApp",NaN,NaN,NaN
4,81zo6pr,7YE0RR,2025-04-22 15:40:00,RECTJNG0MJ0DJP1KV,1.0,NaN,NaN,No,No,Yes,Yes,Not satisfied with my phone,Yes,"Yes, but I am not satisfied with the MoApp",MoPhones App,Yes,Not satisfied with my phone
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4124,7R0XpWz,KYxJ2l7,2025-12-22 06:27:00,RECOWEFSLZI54IBO7,10.0,It is awesome,Allow installation from other sources...admin ...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4125,q4QdM7Y,pbOxobJ,2025-12-23 19:44:00,REC0NSEYY825H22NP,0.0,You normally don't solve clients issues you ju...,Those who are in charge of repairs to stop laz...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4126,aQJZ6qX,EkWdgoL,2025-12-26 04:17:00,RECDLIUM5PL83CCCH,4.0,Sometimes my mpesa doesn't work,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4127,QK8RME8,BPOMzK,2025-12-26 21:42:00,RECLNYF4D3HLACA2X,8.0,Customer service,Customer service takes,Yes,Yes,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### 🆔 Identity / Structure Variables

These columns identify the loan and define the time context of each record.

*   **LOAN_ID** → Unique identifier for each loan (key column)
*   **DATE** → Reporting (snapshot) date on which all loan metrics are calculated
*   **SALE_DATE** → Loan origination date (when the device was sold)

### 💰 Payment Behavior (Core Risk Signals)

These variables describe how customers are repaying their loans over time.

*   **TOTAL_PAID** → Cumulative amount paid by the customer up to the snapshot date
*   **TOTAL_DUE_TODAY** → Total amount the customer is expected to have paid by this date
*   **DEPOSIT** → at loan start (structural / contract feature)
*   **ADVANCE** → Amount the customer has paid ahead of schedule (early payments after loan start)  → during loan lifecycle (behavioral / repayment feature)

### 📝 Contract / Pricing Variables

These describe the structure of the loan agreement.

*   **WEEKLY_RATE** → Scheduled weekly repayment amount for the loan
*   **INITIAL_PAY** → First payment made at the start of the loan
*   **EXPECTED_PAYMENT** → Payment amount expected for the reporting period
*   **DISCOUNT** → Discount applied to the loan or cash price
*   **CREDIT_CHECK_DONE** → Result or status of the credit check performed

### Columns to DROP (as they reveal the result or are problematic)

These columns are identified as potentially problematic, either because they directly reveal information that a predictive model should ideally infer (data leakage), or because they have too many missing values to be useful.

*   **CLOSING_BALANCE**
    *   **Reason:** Shows the final loan state, which can lead to data leakage.
*   **BALANCE**
    *   **Reason:** Can indirectly reveal whether the customer paid well or not, making it difficult to predict for new customers.
*   **BALANCE_DUE_STATUS**
    *   **Reason:** Already indicates if a customer is behind on payments.
*   **DAYS_PAST_DUE**
    *   **Reason:** Directly shows how late the customer is.
    *   **Decision Rationale:** Both `BALANCE_DUE_STATUS` and `DAYS_PAST_DUE` inherently contain the outcome we might want to predict (customer risk), making them unsuitable as features for a forward-looking model.
*   **ARREARS**
    *   **Reason:** Directly shows the unpaid amount, similar to the above points.
*   **PAYMENT_AMOUNT**
    *   **Reason:** Almost all values are missing (~99.8% missing, 30 non-null out of 20,742), providing no signal and just noise.
*   **ADJUSTMENT_AMOUNT**
    *   **Reason:** Almost all values are missing, similar to `PAYMENT_AMOUNT`.
*   **PREPAYMENT_AMOUNT**
    *   **Reason:** While it represents early/upfront payment, `INITIAL_PAY` (money paid at the beginning of the loan) is already kept. Keeping both is redundant for upfront payments.
*   **OVERPAYMENT_AMOUNT**
    *   **Reason:** About 97% of rows are zero, creating highly unbalanced/skewed data. It should be used only as a small secondary signal, not as a major feature.
*   **EXPECTED_PAYMENT**
    *   **Reason:** Many rows are missing or zero (~88%). However, when values exist, they contain useful information. Should be treated carefully if kept.
*   **BALANCE_DUE_TO_DATE**
    *   **Reason:** Directly indicates if the customer is ahead or behind on payments. It is a strong behavioral/performance signal that is very close to the final outcome (good or bad customer).
*   **CREDIT_EXPIRY**
    *   **Reason:** Does not directly tell how the customer behaves, pays, or if they are risky. It is a contractual date, not a behavioral indicator.
*   **TOTAL_PAID_WITH_ADJUSTMENTS_15D**
    *   **Reason:** If the goal is to predict if a customer will become risky, this column already reveals how well they performed in the early critical window, leading to data leakage. The model would learn the outcome directly from this feature.
*   **RETURN_DATE**
    *   **Reason:** Too many missing values (91%).

### FEATURE ENGINEERING

#### 1. Target Variables (`engineer_target_variables`)
These features are different definitions of 'default' or payment performance, allowing for flexibility in modeling objectives.

*   `TARGET_DEFAULT_BINARY`: A binary flag (0 or 1) indicating if the account status is 'PAR 7', 'PAR 30', or 'Inactive' based on `ACCOUNT_STATUS_L2`.
*   `TARGET_SEVERE_DEFAULT`: A binary flag (0 or 1) indicating if the account status is 'PAR 30' (severe delinquency) based on `ACCOUNT_STATUS_L2`.
*   `TARGET_FPD`: A binary flag (0 or 1) indicating if the loan has a 'First Payment Default' based on `ACCOUNT_STATUS_L1`.
*   `TARGET_DPD`: The number of `DAYS_PAST_DUE`, representing a continuous target for regression tasks (filled with 0 for non-missing values).

In [34]:
# TARGET VARIABLE ENGINEERING

def engineer_target_variables(credit_df):
    """
    Create multiple target definitions for default prediction.
    Returns dataframe with target columns.
    """
    df = credit_df.copy()

    # Target 1: Binary default based on ACCOUNT_STATUS_L2
    # PAR = Past Due / Problematic accounts
    df['TARGET_DEFAULT_BINARY'] = df['ACCOUNT_STATUS_L2'].isin(['PAR 7', 'PAR 30', 'Inactive']).astype(int)

    # Target 2: Severe default (PAR 30+ - serious delinquency)
    df['TARGET_SEVERE_DEFAULT'] = df['ACCOUNT_STATUS_L2'].isin(['PAR 30']).astype(int)

    # Target 3: Early-stage delinquency (FPD - First Payment Default)
    df['TARGET_FPD'] = df['ACCOUNT_STATUS_L1'].str.contains('FPD', na=False).astype(int)

    # Target 4: Days past due as continuous target (regression)
    df['TARGET_DPD'] = df['DAYS_PAST_DUE'].fillna(0)

    return df

credit = engineer_target_variables(clean_credit_snapshot)



#### 2. Payment Behavior Features (`engineer_payment_behavior_features`)
These features capture aspects of how customers make payments, including ratios, gaps, and consistency.

*   `PAYMENT_RATIO`: The ratio of `TOTAL_PAID` to `TOTAL_DUE_TODAY`, indicating payment fulfillment.
*   `PAYMENT_GAP`: The difference between `TOTAL_DUE_TODAY` and `TOTAL_PAID`, showing how much is unpaid or overpaid.
*   `PAYMENT_ADVANCE_RATIO`: The ratio of `ADVANCE` amount to `TOTAL_DUE_TODAY`, reflecting proactive payments.
*   `ARREARS_TO_DUE_RATIO`: The ratio of `ARREARS` to `TOTAL_DUE_TODAY`, indicating the severity of outstanding arrears.
*   `ARREARS_TO_BALANCE_RATIO`: The ratio of `ARREARS` to `BALANCE`, showing arrears relative to the current outstanding balance.
*   `PAYMENT_MADE`: A binary flag (0 or 1) indicating if a `PAYMENT` was made on the snapshot date.
*   `CUMULATIVE_PAYMENTS`: The cumulative count of payments made over the loan's history up to the snapshot date.
*   `TOTAL_EXPECTED_PAYMENTS`: The cumulative sum of `EXPECTED_PAYMENT` over the loan's history.
*   `PAYMENT_CONSISTENCY_SCORE`: A score reflecting the consistency of payments, calculated as `CUMULATIVE_PAYMENTS` divided by the number of snapshots for the loan.
*   `ARREARS_CHANGE`: The daily change in `ARREARS`, indicating whether arrears are increasing or decreasing.
*   `ARREARS_INCREASING`: A binary flag (0 or 1) if `ARREARS_CHANGE` is positive, indicating worsening arrears.


In [35]:
# ============================================
# FEATURE ENGINEERING
# ============================================

def engineer_payment_behavior_features(credit_df):
    """Features capturing customer payment behavior."""
    df = credit_df.copy()

    # 1. Payment ratio metrics
    df['PAYMENT_RATIO'] = df['TOTAL_PAID'] / (df['TOTAL_DUE_TODAY'] + 1)
    df['PAYMENT_GAP'] = df['TOTAL_DUE_TODAY'] - df['TOTAL_PAID']
    df['PAYMENT_ADVANCE_RATIO'] = df['ADVANCE'] / (df['TOTAL_DUE_TODAY'] + 1)

    # 2. Arrears severity
    df['ARREARS_TO_DUE_RATIO'] = df['ARREARS'] / (df['TOTAL_DUE_TODAY'] + 1)
    df['ARREARS_TO_BALANCE_RATIO'] = df['ARREARS'] / (df['BALANCE'] + 1)

    # 3. Payment consistency (rolling metrics per loan)
    df = df.sort_values(['LOAN_ID', 'DATE'])
    df['PAYMENT_MADE'] = (df['PAYMENT'] > 0).astype(int)

    # Cumulative payment consistency
    df['CUMULATIVE_PAYMENTS'] = df.groupby('LOAN_ID')['PAYMENT_MADE'].cumsum()
    df['TOTAL_EXPECTED_PAYMENTS'] = df.groupby('LOAN_ID')['EXPECTED_PAYMENT'].cumsum()
    df['PAYMENT_CONSISTENCY_SCORE'] = df['CUMULATIVE_PAYMENTS'] / (df.groupby('LOAN_ID').cumcount() + 1)


    # 5. Rolling arrears trend (increasing or decreasing)
    df['ARREARS_CHANGE'] = df.groupby('LOAN_ID')['ARREARS'].diff()
    df['ARREARS_INCREASING'] = (df['ARREARS_CHANGE'] > 0).astype(int)

    return df

#### 3. Temporal Features (`engineer_temporal_features`)
These features derive information from dates, capturing loan age, remaining term, and seasonality.


*   `DAYS_TO_EXPIRY`: The number of days remaining until `CREDIT_EXPIRY`.
*   `DAYS_PAST_NEXT_INVOICE`: The number of days `DATE` is past `NEXT_INVOICE_DATE`.
*   `IS_PAYMENT_LATE`: Binary flag (0 or 1) if `DAYS_PAST_NEXT_INVOICE` is greater than 0.


In [42]:
def engineer_temporal_features(credit_df):
    """Features based on time and loan age."""
    df = credit_df.copy()

    # 1. Remaining term
    df['DAYS_TO_EXPIRY'] = (df['CREDIT_EXPIRY'] - df['DATE']).dt.days

    # 2. Payment timing relative to due date
    df['DAYS_PAST_NEXT_INVOICE'] = (df['DATE'] - df['NEXT_INVOICE_DATE']).dt.days
    df['IS_PAYMENT_LATE'] = (df['DAYS_PAST_NEXT_INVOICE'] > 0).astype(int)

    return df

#### 4. Financial Ratios (`engineer_financial_ratios`)
These features focus on financial health, leverage, and specific payment behaviors related to initial and regular payments.

*   `INITIAL_PAY_RATIO`: Ratio of `INITIAL_PAY` to `WEEKLY_RATE`, showing initial commitment relative to regular payments.
*   `DEPOSIT_RATIO`: Ratio of `DEPOSIT` to `WEEKLY_RATE`.
*   `TOTAL_OBLIGATION`: An approximation of the annual obligation based on `WEEKLY_RATE`.
*   `HAS_OVERPAYMENT`: Binary flag (0 or 1) if `OVERPAYMENT_AMOUNT` is greater than 0.
*   `BALANCE_TO_OBLIGATION`: Ratio of `BALANCE` to `WEEKLY_RATE`, indicating outstanding balance relative to weekly payment.
*   `BALANCE_REDUCTION_RATE`: The rate at which the `BALANCE` is being reduced over time.
*   `NET_PAYMENT_POSITION`: The difference between `ADVANCE` and `ARREARS`, indicating net financial standing.
*   `POSITION_STATUS`: Categorical feature ('Ahead', 'Behind', 'OnTrack') based on `NET_PAYMENT_POSITION`.

In [37]:
def engineer_financial_ratios(credit_df):
    """Features capturing financial health and leverage."""
    df = credit_df.copy()

    # 1. Initial payment effectiveness
    df['INITIAL_PAY_RATIO'] = df['INITIAL_PAY'] / (df['WEEKLY_RATE'] + 1)
    df['DEPOSIT_RATIO'] = df['DEPOSIT'] / (df['WEEKLY_RATE'] + 1)

    # 2. Loan structure metrics
    df['TOTAL_OBLIGATION'] = df['WEEKLY_RATE'] * 52  # Approximate annual obligation

    # 3. Overpayment patterns
    df['HAS_OVERPAYMENT'] = (df['OVERPAYMENT_AMOUNT'] > 0).astype(int)

    # 4. Balance management
    df['BALANCE_TO_OBLIGATION'] = df['BALANCE'] / (df['WEEKLY_RATE'] + 1)
    df['BALANCE_REDUCTION_RATE'] = -df.groupby('LOAN_ID')['BALANCE'].diff() / (df['BALANCE'] + 1)

    # 5. Advanced vs Arrears net position
    df['NET_PAYMENT_POSITION'] = df['ADVANCE'] - df['ARREARS']
    df['POSITION_STATUS'] = np.where(df['NET_PAYMENT_POSITION'] > 0, 'Ahead',
                                     np.where(df['NET_PAYMENT_POSITION'] < 0, 'Behind', 'OnTrack'))

    return df

#### 5. Risk Scores (`engineer_risk_scores`)
These are composite scores designed to aggregate multiple signals into a single risk indicator.

*   `PAYMENT_BEHAVIOR_SCORE`: A composite score (0-100) combining `PAYMENT_RATIO`, `ARREARS_TO_DUE_RATIO`, `ADVANCE` flag, and `PAYMENT_CONSISTENCY_SCORE`.
*   `EARLY_WARNING_SCORE`: A score (0-100) combining `DAYS_PAST_DUE` flag, `ARREARS_INCREASING` flag, and `PAYMENT_RATIO` below a threshold, designed to detect early signs of trouble.
*   `CUSTOMER_STABILITY_SCORE`: A score (0-100) combining `CREDIT_CHECK_DONE` status, `PAYMENT_CONSISTENCY_SCORE`, and `CUSTOMER_AGE`.
*   `RISK_TIER`: Categorical feature ('High Risk', 'Medium Risk', 'Low Risk', 'Minimal Risk') based on `EARLY_WARNING_SCORE`.


In [38]:
def engineer_risk_scores(credit_df):
    """Composite risk scores from multiple signals."""
    df = credit_df.copy()

    # 1. Payment behavior score (0-100, higher = better)
    df['PAYMENT_BEHAVIOR_SCORE'] = np.clip(
        100 * (
            (df['PAYMENT_RATIO'] * 0.4) +
            (1 - np.minimum(1, df['ARREARS_TO_DUE_RATIO'])) * 0.3 +
            (df['ADVANCE'] > 0).astype(int) * 0.2 +
            df['PAYMENT_CONSISTENCY_SCORE'] * 0.1
        ), 0, 100
    )

    # 2. Early warning score (detects potential default before it happens)
    df['EARLY_WARNING_SCORE'] = np.clip(
        (
            (df['DAYS_PAST_DUE'] > 0).astype(int) * 30 +
            (df['ARREARS_INCREASING'] == 1).astype(int) * 25 +
            (df['PAYMENT_RATIO'] < 0.8).astype(int) * 20
        ), 0, 100
    )

    # 3. Customer stability score
    df['CUSTOMER_STABILITY_SCORE'] = np.clip(
        (
            (df['CREDIT_CHECK_DONE'] == 'Pass').astype(int) * 40 +
            (df['PAYMENT_CONSISTENCY_SCORE'] * 30) +
            ((df['CUSTOMER_AGE'] / 365) * 30)
        ), 0, 100
    )

    # 4. Delinquency risk tier
    df['RISK_TIER'] = np.where(
        df['EARLY_WARNING_SCORE'] >= 70, 'High Risk',
        np.where(df['EARLY_WARNING_SCORE'] >= 40, 'Medium Risk',
                 np.where(df['EARLY_WARNING_SCORE'] >= 20, 'Low Risk', 'Minimal Risk'))
    )

    return df

#### 6. First Payment Features (`engineer_first_payment_features`)
These features are specifically designed to capture behavior around the initial stages of the loan, particularly for First Payment Default (FPD).

*   `FIRST_PAYMENT_RATIO`: Ratio of `FIRST_PAYMENT` to `FIRST_EXPECTED_PAYMENT`.
*   `FIRST_PAYMENT_SHORTFALL`: The difference between `FIRST_EXPECTED_PAYMENT` and `FIRST_PAYMENT` if `FIRST_EXPECTED_PAYMENT` is greater.
*   `EARLY_15D_PAYMENT_RATIO`: Ratio of `TOTAL_PAID_WITH_ADJUSTMENTS_15D` to an approximation of two weeks of `WEEKLY_RATE`.
*   `EARLY_15D_ADEQUATE`: Binary flag (0 or 1) if `EARLY_15D_PAYMENT_RATIO` is 0.5 or more.
*   `INITIAL_DEPOSIT_GAP`: The difference between `INITIAL_PAY` and `DEPOSIT`.
*   `INITIAL_DEPOSIT_RATIO`: Ratio of `INITIAL_PAY` to `DEPOSIT`.

In [45]:
def engineer_first_payment_features(credit_df):
    """First Payment Default (FPD) specific features."""
    df = credit_df.copy()

    # 1. First payment performance
    df['FIRST_PAYMENT_RATIO'] = df['FIRST_PAYMENT'] / (df['FIRST_EXPECTED_PAYMENT'] + 1)
    df['FIRST_PAYMENT_SHORTFALL'] = np.maximum(0, df['FIRST_EXPECTED_PAYMENT'] - df['FIRST_PAYMENT'])

    # 2. Early payment behavior (first 15 days)
    df['EARLY_15D_PAYMENT_RATIO'] = df['TOTAL_PAID_WITH_ADJUSTMENTS_15D'] / (df['WEEKLY_RATE'] * 2 + 1)
    df['EARLY_15D_ADEQUATE'] = (df['EARLY_15D_PAYMENT_RATIO'] >= 0.5).astype(int)

    # 3. Initial pay vs deposit differential
    df['INITIAL_DEPOSIT_GAP'] = df['INITIAL_PAY'] - df['DEPOSIT']
    df['INITIAL_DEPOSIT_RATIO'] = df['INITIAL_PAY'] / (df['DEPOSIT'] + 1)

    return df

#### 8. Credit Check Features (`engineer_credit_check_features`)
These features encode information from the credit check process.

*   `CREDIT_CHECK_ENCODED`: Numerical encoding of `CREDIT_CHECK_DONE` (Pass=2, Pending=1, Fail/Not Done=0).
*   `NO_CREDIT_CHECK`: Binary flag (0 or 1) if a credit check was not 'Pass' or 'Pending'.

In [43]:
def engineer_credit_check_features(credit_df):
    """Features based on credit check results."""
    df = credit_df.copy()

    # Encode credit check results
    credit_check_map = {
        'Pass': 2,
        'Pending': 1,
        'Fail': 0,
        'Not Done': 0
    }
    df['CREDIT_CHECK_ENCODED'] = df['CREDIT_CHECK_DONE'].map(credit_check_map).fillna(0)

    # Flag for missing/insufficient credit check
    df['NO_CREDIT_CHECK'] = (~df['CREDIT_CHECK_DONE'].isin(['Pass', 'Pending'])).astype(int)

    return df

In [46]:
# ============================================
# APPLY ALL FEATURE ENGINEERING
# ============================================

def engineer_all_features(credit_df):
    """Apply all feature engineering functions."""
    print("Starting feature engineering...")

    df = credit_df.copy()

    # Store original columns to identify new features
    original_cols = set(df.columns)

    # Apply all feature groups
    df = engineer_target_variables(df)
    df = engineer_payment_behavior_features(df)
    df = engineer_temporal_features(df)
    df = engineer_financial_ratios(df)
    df = engineer_risk_scores(df)
    df = engineer_first_payment_features(df)
    df = engineer_credit_check_features(df)

    # Identify new features
    new_cols = set(df.columns) - original_cols
    print(f"\nCreated {len(new_cols)} new features")
    print("\nNew features:")
    for col in sorted(new_cols):
        print(f"  - {col}")

    return df

# Apply all feature engineering
credit_engineered = engineer_all_features(credit)

Starting feature engineering...

Created 34 new features

New features:
  - ARREARS_CHANGE
  - ARREARS_INCREASING
  - ARREARS_TO_BALANCE_RATIO
  - ARREARS_TO_DUE_RATIO
  - BALANCE_REDUCTION_RATE
  - BALANCE_TO_OBLIGATION
  - CREDIT_CHECK_ENCODED
  - CUMULATIVE_PAYMENTS
  - CUSTOMER_STABILITY_SCORE
  - DAYS_PAST_NEXT_INVOICE
  - DAYS_TO_EXPIRY
  - DEPOSIT_RATIO
  - EARLY_15D_ADEQUATE
  - EARLY_15D_PAYMENT_RATIO
  - EARLY_WARNING_SCORE
  - FIRST_PAYMENT_RATIO
  - FIRST_PAYMENT_SHORTFALL
  - HAS_OVERPAYMENT
  - INITIAL_DEPOSIT_GAP
  - INITIAL_DEPOSIT_RATIO
  - INITIAL_PAY_RATIO
  - IS_PAYMENT_LATE
  - NET_PAYMENT_POSITION
  - NO_CREDIT_CHECK
  - PAYMENT_ADVANCE_RATIO
  - PAYMENT_BEHAVIOR_SCORE
  - PAYMENT_CONSISTENCY_SCORE
  - PAYMENT_GAP
  - PAYMENT_MADE
  - PAYMENT_RATIO
  - POSITION_STATUS
  - RISK_TIER
  - TOTAL_EXPECTED_PAYMENTS
  - TOTAL_OBLIGATION


In [48]:
cat_cols = credit_engineered.select_dtypes('object').columns.tolist()
cat_cols.remove('LOAN_ID')
for col in cat_cols:
  print(f'----- {col} VALUES -----')
  print(f'{credit_engineered[col].unique()}\n')

----- CREDIT_CHECK_DONE VALUES -----
['Y LOAN' 'Y CASH' 'Z DEMO' 'Y LOST']

----- ACCOUNT_STATUS_L1 VALUES -----
['Paid Off Loan' 'New Active' 'Inactive 15-30' 'Write Off'
 'First Month Default without inventory over 30' 'Inactive 08-14'
 'Blocked 31-60' 'First Payment Default' 'Existing Active'
 'Inactive 01-07' 'Blocked 61-90' 'Paid Off Cash'
 'Cancelled Returned past 3 mths' 'First 2 days Return'
 'First Month Default without inventory 01-07'
 'First Month Default with inventory' 'Demo' 'Lost Write Off'
 'Cancelled Returned in 3 mths'
 'First Month Default without inventory 08-14']

----- ACCOUNT_STATUS_L2 VALUES -----
['Paid Off' 'Active' 'PAR 7' 'PAR 30' 'FMD' 'FPD' 'Inactive' 'Return'
 'Unknown']

----- POSITION_STATUS VALUES -----
['Behind' 'Ahead' 'OnTrack']

----- RISK_TIER VALUES -----
['Minimal Risk' 'Medium Risk' 'High Risk' 'Low Risk']

